pkg/controllers/metrics/pod/counter.go

```diff
@@ -0,0 +1,131 @@
+*/
+
+package pod
+
+import (
+	"strings"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha4"
+	"github.com/awslabs/karpenter/pkg/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	metricNamespace = metrics.KarpenterNamespace
+	metricSubsystem = "pods"
+
+	metricLabelPhase       = "phase"
+	metricLabelProvisioner = metrics.ProvisionerLabel
+	metricLabelZone        = "zone"
+
+	nodeLabelZone = v1.LabelTopologyZone
+)
+
+var (
+	knownValuesForNodeLabels = v1alpha4.WellKnownLabels
+	phaseValues              = []v1.PodPhase{
+		v1.PodFailed,
+		v1.PodPending,
+		v1.PodRunning,
+		v1.PodSucceeded,
+		v1.PodUnknown,
+	}
+
+	provisionablePodCount = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "provisionable_count",
+			Help:      "Total count of pods that are eligible for provisioning.",
+		},
+	)
+
+	podCountByPhaseProvisioner = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "count",
+			Help:      "Total pod count by phase and provisioner.",
+		},
+		[]string{
+			metricLabelPhase,
+			metricLabelProvisioner,
+		},
+	)
+
+	runningPodCountByProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "running_count",
+			Help:      "Total running pod count by provisioner and zone.",
+		},
+		[]string{
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(provisionablePodCount)
+	crmetrics.Registry.MustRegister(podCountByPhaseProvisioner)
+	crmetrics.Registry.MustRegister(runningPodCountByProvisionerZone)
+}
+
+func publishPodCounts(provisioner string, podsByZone map[string][]v1.Pod, provisionablePods []*v1.Pod) error {
+	countByPhase := make(map[v1.PodPhase]int, len(phaseValues))
+	zoneValues := knownValuesForNodeLabels[nodeLabelZone]
+	countByZone := make(map[string]int, len(zoneValues))
+
+	for zone, pods := range podsByZone {
+		countByZone[zone] = len(pods)
+
+		for _, pod := range pods {
+			countByPhase[pod.Status.Phase] += 1
+		}
+	}
+
+	errors := make([]error, 0, len(phaseValues)+len(zoneValues))
+
+	provisionablePodCount.Set(float64(len(provisionablePods)))
+	for _, phase := range phaseValues {
+		metricLabels := prometheus.Labels{
+			metricLabelPhase:       strings.ToLower(string(phase)),
+			metricLabelProvisioner: provisioner,
+		}
+		errors = append(errors, publishCount(podCountByPhaseProvisioner, metricLabels, countByPhase[phase]))
+	}
+	for _, zone := range zoneValues {
+		metricLabels := prometheus.Labels{
+			metricLabelProvisioner: provisioner,
+			metricLabelZone:        zone,
+		}
+		errors = append(errors, publishCount(runningPodCountByProvisionerZone, metricLabels, countByZone[zone]))
```

Comment:
check out multierr.Append instead of managing the slice of errors yourself.

Reply:
Hopefully they wouldn't all error. The typical pattern I've seen used across the k8s space (and in this project) is:
```
if err := publicCount(...); err != nil {
  errs = multierr.Append(errs, err)
}
```